### PR TITLE
Use principalName for creating user name

### PR DESCRIPTION
**Pull request by dramich**
Problem:
There is a race when creating users which can cause duplicate users
for the same principal to get created

Solution:
Use a deterministic name for the user based off the principalName so k8s
can tell us if that user already exists

Issue: https://github.com/rancher/rancher/issues/14822

### CODE REVIEW

**Comment by vincent99** [comment]
You have 10 characters of a hexadecimal string, which is 40 bits of info or ~1.1 trillion possible IDs.  That sounds like way more than you'd ever need, but because of the "birthday paradox" it's really on the lowish end of "probably good enough that some user won't hit it this year".

It doesn't take all that many users before there's a non-negligible chance of the next new user hashing to the same 10 characters as an existing (legitimately different) user.  With 1 user it's 1 in 1.1 trillion.  At ~1,500 user entries, it's "one in a million".  At 50,000 users (a large company, or a user creation script gone bad for a while), you're down to less than one in a thousand.

So I would say the ID should contain more bits... for example, there's already a base32 library loaded here.  10 base32 characters is 50 bits instead of 40, so that's 1024x the address space and makes 50,000 users "one in a million".

File: pkg/auth/providers/common/usermanager.go [hunk 1]
```diff
@@ -265,16 +266,23 @@ func (m *userManager) EnsureUser(principalName, displayName string) (*v3.User, e
 		// User doesn't exist, create user
 		logrus.Infof("Creating user for principal %v", principalName)
 
+		// Create a hash of the principalName to use as the name for the user,
+		// this lets k8s tell us if there are duplicate users with the same name
+		// thus avoiding a race.
+		hasher := sha256.New()
+		hasher.Write([]byte(principalName))
+		sha := base32.StdEncoding.WithPadding(-1).EncodeToString(hasher.Sum(nil))[:10]
```
Comment (cjellick) [inline pkg/auth/providers/common/usermanager.go]: will this panic if length of the encoding string is less than 10? if so, can we protect against that?

Comment (dramich) [reply pkg/auth/providers/common/usermanager.go]: Technically, yes, but sha256 always returns the same length of byte array so we are always encoding to the same length of string (52) regardless of the length of the input string.

**Comment by cjellick** [comment]
LGTM

**Pull request merged**